File: projects/07/app/src/main/java/io/github/ironking24/nand2tetris/VMTranslator/CodeWriter.java
package io.github.ironking24.nand2tetris.VMTranslator;

import javax.naming.LimitExceededException;
import java.io.BufferedWriter;
import java.io.Closeable;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;

/**
 * this class is responsible for translating VM code into assembly instructions and then writing them into the .asm file.
*/
public class CodeWriter implements Closeable{

    private final BufferedWriter writer;
    private final ArrayList<String> statics = new ArrayList<>();
    private final HashMap<String, String> symbolTable = new HashMap<>(9);
    private final HashMap<String, Integer> runningNumbersTable = new HashMap<>(2);

    private String currentFileName;

    CodeWriter(BufferedWriter output) throws IOException{
        this.writer = output;
        this.currentFileName = "";
        
        //initialize vm to asm translation
        symbolTable.put("add","+");
        symbolTable.put("sub","-");
        symbolTable.put("neg","-");
        symbolTable.put("not","!");
        symbolTable.put("and","&");
        symbolTable.put("or","|");
        symbolTable.put("eq","JEQ");
        symbolTable.put("gt","JGT");
        symbolTable.put("lt","JLT");
        symbolTable.put("argument","ARG");
        symbolTable.put("local","LCL");
        symbolTable.put("this","THIS");
        symbolTable.put("that","THAT");

        //BootStrap
        writer.write(
            "@256\n"+
            "D=A\n"+
            "@SP\n"+
            "M=D\n"
            );
    }

    /**
     * sets the file name of the current file being translated.
    */
    public void setFileName(String name){
        currentFileName = name;
    }
    
    /**
     * writes the operation execution on the stack in hack's assembly.
     * @throws IllegalArgumentException if the @param command is not recognized.
    */
    public void writeArithmetic(String command) throws Exception{
        switch(command){
            case "add": case "sub":
            case "and": case "or":
                writer.write(String.format(
                "@SP\n"+
                "AM=M-1\n"+
                "D=M\n"+
                "A=A-1\n"+
                "M=M%sD\n", 
                symbolTable.get(command)));
                break;
            case "neg": case "not":
                writer.write(String.format(
                "@SP\n"+
                "A=M-1\n"+
                "M=%sM\n", 
                symbolTable.get(command)));
                break;
            case "eq": case "gt": case"lt":
                String checkName = String.format("%s$Check%d", currentFileName, nextRunningNumber(String.format("%s$Check",currentFileName)));
                writer.write(String.format(
                "@SP\n"+
                "AM=M-1\n"+
                "D=M\n"+
                "@SP\n"+
                "AM=M-1\n"+
                "D=M-D\n"+
                "M=-1\n"+
                "@%s\n"+
                "D;%s\n"+
                "@SP\n"+
                "A=M\n"+
                "M=0\n"+
                "(%s)\n"+
                "@SP\n"+
                "AM=M+1\n", 
                checkName, symbolTable.get(command), checkName));
                break;
            default:
                throw new IllegalArgumentException(String.format("Unknown arithmetic command %s.", command));
        }
    }

    /**
     * writes the push/pop execution on the stack in hack's assembly.
     * @throws IllegalArgumentException is thrown when the @param command operation, @param segment or the static is unrecognized.
     * @throws IOException is thrown when I/O error occurs.
     * @throws LimitExceededException is thrown when there is no more room for new statics
     * @throws IndexOutOfBoundsException is thrown when the @param index is exceeds the segment limits
    */
    public void writePushPop(Parser.commandTypes command, String segment, int index) throws IllegalArgumentException, IOException, LimitExceededException, IndexOutOfBoundsException{
        switch(command){
            case C_PUSH:
                switch(segment){
                    case "constant":
                        if(index < -32767 || index > 32767){
                            throw new IndexOutOfBoundsException(String.format("Illegal constant value: %s", index));
                        }
                        writer.write(String.format(
                            "@%d\n"+
                            "D=A\n", 
                            index));
                        break;
                    case "local": case "argument": case "this": case "that":
                        writer.write(String.format("@%s\n", symbolTable.get(segment)));
                        if (index < 0 || index > 32767){
                            throw new IndexOutOfBoundsException(String.format("Illegal segment(%s) index is out of bounds: %s", segment, index));
                        }
                        else if(index > 0){
                            writer.write(String.format(
                            "D=M\n"+
                            "@%d\n"+
                            "A=D+A\n",
                            index));
                            }
                        else{
                            writer.write("A=M\n");
                        }
                        writer.write("D=M\n");
                        break;
                    case "pointer":
                        if(!(index >= 0 && index < 2))
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the pointer segment, the index is out of bounds: %d", index));
                        }
                        writer.write(String.format(
                        "@R%d\n"+
                        "D=M\n",
                        index + 3));
                        break;
                    case "temp":
                        if(!(index >= 0 && index <= 7))
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the temp segment, index is out of bounds: %d", index));
                        }
                        writer.write(String.format(
                        "@R%d\n"+
                        "D=M\n",
                        index + 5));
                        break;
                    case "static":
                        if(!statics.contains(String.format("%s.%d", currentFileName, index)))
                        {
                            throw new IllegalArgumentException(String.format("the static address for the push operation (%s.%d) does not exist.", currentFileName, index));
                        }
                        writer.write(String.format(
                        "@%s.%d\n"+
                        "D=M\n"
                        , currentFileName, index));
                        break;
                    default:
                        throw new IllegalArgumentException(String.format("Illegal push segment: %s", segment));
                }
                writer.write(
                "@SP\n"+
                "A=M\n"+
                "M=D\n"+
                "@SP\n"+
                "M=M+1\n");
                break;
            case C_POP:
                switch(segment){
                    case "local": case "argument": case "this": case "that":
                        if(index < 0 || index > 32767){
                            throw new IndexOutOfBoundsException(String.format("Illegal segment(%s) index is out of bounds: %s", segment, index));
                        }
                        writer.write(String.format(
                        "@%s\n"+
                        "D=M\n",
                            symbolTable.get(segment)));
                        if(index > 0){
                            writer.write(String.format(
                            "@%d\n"+
                            "D=D+A\n",
                            index));
                        }
                        writer.write(
                        "@R13\n"+
                        "M=D\n"+
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@R13\n"+
                        "A=M\n"+
                        "M=D\n");
                        break;
                    case "pointer":
                        if(index < 0 || index > 2)
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the pointer segment, index is out of bounds: %s", index));
                        }
                        writer.write(String.format(
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@R%d\n"+
                        "M=D\n",
                        index + 3));
                        break;
                    case"temp":
                        if(index < 0 || index > 7)
                        {
                            throw new IndexOutOfBoundsException(String.format("Illegal push to the temp segment, index is out of bounds: %s", index));
                        }
                        writer.write(String.format(
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@R%d\n"+
                        "M=D\n",
                        index + 5));
                        break;
                    case "static":
                        if(!statics.contains(String.format("%s.%d", currentFileName, index)))
                        {
                            if(statics.size()+1 > 240){
                                throw new LimitExceededException("The max number of statics has been exceeded.");
                            }
                            statics.add(String.format(
                            "%s.%d",
                            currentFileName, index));
                        }
                        writer.write(String.format(
                        "@SP\n"+
                        "AM=M-1\n"+
                        "D=M\n"+
                        "@%s.%d\n"+
                        "M=D\n",
                        currentFileName, index));
                        break;
                    default:
                        throw new IllegalArgumentException(String.format("Illegal pop segment: %s", segment));
                }
                break;
            default:
                throw new IllegalArgumentException(String.format("Illegal push/pop command: %s", command));
        }
    }

    /**
     * Closes the BufferedWriter.
     * @throws IOException raised when there is a problem closing the writing stream.
    */
    @Override
    public void close() throws IOException{
        writer.write(
        "(end)\n"+
        "@end\n"+
        "0;JMP\n");
        writer.flush();
        writer.close();
    }

    public void comment(String text) throws IOException{
        writer.write(String.format("//%s\n", text));
    }

    private int nextRunningNumber(String label){
        if(runningNumbersTable.containsKey(label)){
            return runningNumbersTable.merge(label, 1, Integer::sum);
        }
        else{
            runningNumbersTable.put(label, 0);
            return 0;
        }
    }
}

File: projects/07/src/CodeWriter.cpp
#pragma once
#include "CodeWriter.h"
#include <string>

namespace VMTranslator 
{
	CodeWriter::CodeWriter(const std::filesystem::path& output_file): output(output_file, std::ios_base::trunc)
	{
		CodeWriter::output.exceptions(std::ios_base::badbit | std::ios_base::failbit);

		if (!output.is_open() || !output.good()) 
		{
			throw std::runtime_error("CodeWriter failed to file stream.");
		}
	}

	void CodeWriter::writeArithmetic(const std::string& command)
	{
		if (command == "add" || command == "sub" || command == "and" || command == "or")
		{
			output
				<< "@SP\n"
				<< "AM=M-1\n"
				<< "D=M\n" 
				<< "A=A-1\n"
				<< "M=M" << symbol_table.at(command) << "D\n";
		}
		else if (command == "neg" || command == "not")
		{
			output
				<< "@SP\n"
				<< "A=M-1\n"
				<< "M="<< symbol_table.at(command) << "M\n";
		}
		else if (command == "eq" || command == "gt" || command == "lt")
		{
			std::string checkName;

			if (running_number.find(current_fileName + "$Check") == running_number.cend())
			{
				running_number.emplace(current_fileName + "$Check", 1);
				checkName = current_fileName + "$Check0";
			}
			else 
			{
				int val = running_number[current_fileName + "$Check"]++;
				checkName = current_fileName + "$Check" + std::to_string(val);
			}

			output
				<< "@SP\n"
				<< "AM=M-1\n"
				<< "D=M\n"
				<< "@SP\n"
				<< "AM=M-1\n"
				<< "D=M-D\n"
				<< "M=-1\n"
				<< "@" << checkName << '\n'
				<< "D;" << symbol_table.at(command) << '\n'
				<< "@SP\n" 
				<< "A=M\n"
				<< "M=0\n"
				<< "("<< checkName <<")\n"
				<< "@SP\n"
				<< "AM=M+1\n";
		}
	}
	
	void CodeWriter::writePushPop(const Parser::CommandTypes command, const std::string& segment, const short int index)
	{
		switch (command)
		{
		case VMTranslator::Parser::CommandTypes::C_PUSH:
			if (segment == "constant")
			{
				if (index < 0 || index > 32767) {
					throw std::runtime_error("Illegal constant value: " + std::to_string(index));
				}

				output
					<< '@' << index << '\n'
					<< "D=A\n";
			}
			else if (segment == "local" || segment == "argument" || segment == "this" || segment == "that")
			{
				if (index < 0 || index > 32767) {
					throw std::runtime_error("push " + segment + " index is out of bounds : " + std::to_string(index));
				}
				output << "@" << symbol_table.at(segment) << '\n';
				if (index > 0)
				{
					output
						<< "D=M\n"
						<< "@" << index << '\n'
						<< "A=D+A\n";
				}
				else
				{
					output << "A=M\n";
				}
				output << "D=M\n";
			}
			else if (segment == "pointer")
			{
				if (index < 0 || index > 1)
				{
					throw std::runtime_error("Illegal push to the pointer segment, the index is out of bounds: " + std::to_string(index));
				}

				output
					<< "@R" << index + 3 << '\n'
					<< "D=M\n";

			}
			else if (segment == "temp")
			{
				if (index < 0 || index > 7)
				{
					throw std::runtime_error("Illegal push to the temp segment, index is out of bounds: " + std::to_string(index));
				}

				output
					<< "@R" << index + 5 << '\n'
					<< "D=M\n";
			}
			else if ("static")
			{
				std::string label = current_fileName + "." + std::to_string(index);

				if (std::find_if(statics.cbegin(), statics.cend(), [&](std::string s) {return s == label; }) != statics.cend())
				{
					throw std::runtime_error("the static address does not exist for the push operation of: " + label);
				}
				output
					<< "@" << current_fileName << "." << index << '\n'
					<< "D=M\n";
			}
			else
			{
				throw std::runtime_error("Illegal push segment: " + segment);
			}
			output
				<< "@SP\n"
				<< "A=M\n"
				<< "M=D\n"
				<< "@SP\n"
				<< "M=M+1\n";
			break;

		case VMTranslator::Parser::CommandTypes::C_POP:
			if (segment == "local" || segment == "argument" || segment == "this" || segment == "that")
			{
				if (index < 0 || index > 32767) {
					throw std::runtime_error("pop " + segment + " index is out of bounds : " + std::to_string(index));
				}

				output
					<< "@" << symbol_table.at(segment) << '\n'
					<< "D=M\n";

				if (index > 0) 
				{
					output
						<< "@" << index << '\n'
						<< "D=D+A\n";
				}
				output
					<< "@R13\n" 
					<< "M=D\n" 
					<< "@SP\n" 
					<< "AM=M-1\n"
					<< "D=M\n"
					<< "@R13\n"
					<< "A=M\n"
					<< "M=D\n";
				break;
			}
			else if (segment == "pointer")
			{
				if (index < 0 || index > 2)
				{
					throw new std::runtime_error("Illegal push to the pointer segment, index is out of bounds: " + std::to_string(index));
				}
				output
					<< "@SP\n"
					<< "AM=M-1\n"
					<< "D=M\n"
					<< "@R" << index + 3 << '\n'
					<< "M=D\n";
			}
			else if (segment == "temp")
			{
				if (index < 0 || index > 7)
				{
					throw std::runtime_error("Illegal push to the temp segment, index is out of bounds: " + std::to_string(index));
				}
				output
					<< "@SP\n"
					<< "AM=M-1\n"
					<< "D=M\n"
					<< "@R" << index + 5 << '\n'
					<< "M=D\n";
			}
			else if ("static")
			{
				if (std::find(statics.cbegin(), statics.cend(), current_fileName + "." + std::to_string(index))!= statics.cend())
				{
					if (statics.size() + 1 > 240) {
						throw std::runtime_error("The max number of statics has been exceeded.");
					}

					statics.push_back(current_fileName + '.' + std::to_string(index));
				}

				output
					<< "@SP\n"
					<< "AM=M-1\n"
					<< "D=M\n"
					<< '@' << current_fileName << '.' << index << '\n'
					<< "M=D\n";

			}
			else 
			{
				throw std::runtime_error("Illegal pop segment: " + segment);
			}
			break;
		default:
			throw std::runtime_error("Illegal push/pop command at segment: " + segment + " and index " + std::to_string(index));
		}
	}


	void CodeWriter::close()
	{
		output
			<< "(end)\n"
			<< "@end\n"
			<< "0;JMP\n";
		output.flush();
		output.close();
	}

	CodeWriter::~CodeWriter()
	{
		if (output.is_open()) 
		{
			output.flush();
			output.close();
		}
	}
}